go/vt/vttablet/tabletmanager/vreplication/vcopier.go

```diff
@@ -364,3 +631,562 @@ func (vc *vcopier) fastForward(ctx context.Context, copyState map[string]*sqltyp
+	if _, err := vc.vr.setSQLMode(ctx, dbClient); err != nil {
+		return nil, vterrors.Wrap(err, "failed to set sql_mode")
+	}
+	if err := vc.vr.clearFKCheck(dbClient); err != nil {
+		return nil, vterrors.Wrap(err, "failed to clear foreign key check")
+	}
+	return dbClient, nil
+}
+
+func (vc *vcopier) newCopyWorkQueue(
+	workerFactory func(context.Context) (*vcopierCopyWorker, error),
+) *vcopierCopyWorkQueue {
+	concurrent := false
+	maxDepth := 1
+	if isExperimentalParallelBulkInsertsEnabled() {
+		concurrent = true
+		maxDepth = getCopyInsertConcurrency()
+	}
+	return newVCopierCopyWorkQueue(concurrent, maxDepth, workerFactory)
+}
+
+func (vc *vcopier) newCopyWorkerFactory() func(context.Context) (*vcopierCopyWorker, error) {
+	workerFactory := func(_ context.Context) (*vcopierCopyWorker, error) {
+		return newVCopierCopyWorker(
+			false, /* close db client */
+			vc.vr.dbClient,
+		), nil
+	}
+	if isExperimentalParallelBulkInsertsEnabled() {
+		workerFactory = func(ctx context.Context) (*vcopierCopyWorker, error) {
+			dbClient, err := vc.newClientConnection(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create new db client: %s", err.Error())
+			}
+			return newVCopierCopyWorker(
+				true, /* close db client */
+				dbClient,
+			), nil
+		}
+	}
+	return workerFactory
+}
+
+// Enqueue a new copy task. This will obtain a worker from the pool, execute
+// the task with that worker, and afterwards return the worker to the pool. If
+// vcopierCopyWorkQueue is configured to operate concurrently, the task will be
+// executed in a separate goroutine. Otherwise the task will be executed in the
+// calling thread.
+func (vcq *vcopierCopyWorkQueue) Enqueue(ctx context.Context, currT *vcopierCopyTask) error {
+	if !vcq.IsOpen() {
+		return fmt.Errorf("work queue is not open")
+	}
+
+	// Get a handle on an unused worker.
+	poolH, err := vcq.workerPool.Get(ctx, []string{})
+	if err != nil {
+		return fmt.Errorf("failed to get a worker from pool: %s", err.Error())
+	}
+
+	currW := poolH.(*vcopierCopyWorker)
+
+	execute := func(task *vcopierCopyTask) {
+		currW.Execute(ctx, task)
+		vcq.workerPool.Put(poolH)
+	}
+
+	// If the work queue is configured to work concurrently, execute the task
+	// in a separate goroutine. Otherwise execute the task in the calling
+	// goroutine.
+	if vcq.concurrent {
+		go execute(currT)
+	} else {
+		execute(currT)
+	}
+
+	return nil
+}
+
+// Close waits for all workers to be returned to the worker pool.
+func (vcq *vcopierCopyWorkQueue) Close() {
+	if !vcq.IsOpen() {
+		return
+	}
+	vcq.isOpen = false
+	vcq.workerPool.Close()
+}
+
+func (vcq *vcopierCopyWorkQueue) IsOpen() bool {
+	return vcq.isOpen
+}
+
+// Open the work queue. The provided arguments are used to generate INSERT and
+// UPDATE statements.
+func (vcq *vcopierCopyWorkQueue) Open(
+	copyStateUpdate *sqlparser.ParsedQuery,
+	pkfields []*querypb.Field,
+	tablePlan *TablePlan,
+) {
+	if vcq.IsOpen() {
```

Comment:
Any reason to use the public/exported functions within the package?

Reply:
Not sure I understand this q, can you rephrase?